text/0024-schema-syntax.md

```diff
@@ -0,0 +1,405 @@
+```
+"addr": {
+    "ip": { "type": "Extension", "name": "ipaddr" },
+}
+```
+If we _also_ had defined a common type `ipaddr` we would reference it by writing
+```
+"addr": {
+    "myip": { "type": "ipaddr" },
+}
+```
+Common types may also overlap with entity types, for similar reasons: To reference an entity type you must write `{ "type": "Entity", "name": "Foo" }` while to reference a common type you'd just write `{ "type": "Foo" }`.
+
+The custom syntax does not require such designators, for readability. How should type names that are ambiguous therefore be resolved? For example, what should happen when a schema references type `ipaddr` when it also defines `entity ipaddr` and `type ipaddr` ? 
+
+The main proposal already allows using namespace prefixes for disambiguation. If we had entity `ipaddr` and extension type `ipaddr` we could write `My::Namespace::ipaddr` to distinguish the former from the latter, assuming we are using a non-empty namespace. We can do likewise to distinguish entity type `Long` from primitive type `Long`. Namespaces are not sufficient to disambiguate entity types from common types, though, since they both can have namespace prefixes. Moreover, long namespace prefixes could create readability problems.
+
+Here are two possible solutions:
+
+### Typename prefixes
+
+The first solution is to create type-name prefixes for the purpose of disambiguation, which parallel the `"type"` part of the JSON syntax. Proposal:
+1. Extension types can be optionally prefixed with `+`, e.g., `+ipaddr` and `+decimal`
+2. Entity types can be optionally prefixed with `&`, e.g., `&User` and `&My::Namespace::List`.
+
+With these prefixes, we could write the following contrived schema:
+```
+entity ipaddr { name: String };
+type ipaddr = Long;
+entity container {
+    issuer: &ipaddr, // entity ipaddr
+    time: ipaddr,    // type ipaddr
+    addr: +ipaddr    // extension type ipaddr
+};
+```
+You would only write the prefix if there was a possibility for ambiguity. If you leave off a prefix but a name does not uniquely resolve, the schema parser will complain about the ambiguity.
+
+### Forbidding type name overlap
+
+As an alternative solution, we could forbid type-name overlaps entirely. Specifically, we could disallow creating any entity or common types that overlap with extension or primitive types, and we could disallow such types from mutually overlapping within the same namespace.
+
+This solution would lead to better readability (due to fewer concepts imposed on the reader for interpreting type names), but would require imposing a similar restriction on the JSON form of the syntax, which is a breaking change.
+
+As a middle ground, we could choose the first solution above, but _warn_ about overlapping type definitions, to discourage users from defining a common type with the same name as an entity or extension type.
```

Comment:
My preference is for forbidding name overlap, and making the corresponding breaking change in the JSON syntax.

Reply:
I'm in favor of forbidding the name overlap in the case of a conflict between an entity type and a common type in particular.  For conflicts involving extension types it's a little less clear, because then if we implement a new extension type, suddenly some people's existing schemas could become invalid if they happened to use an entity type or common type that has the same name as the new extension type.  This isn't hard to imagine either -- imagine someone writes `type date = int` or something for clarity of attribute types, and then later we actually implement a `date` extension type

Reply:
To summarize your position, @cdisselkoen: for extension types we would use `+` as a prefix for disambiguation but prevent overlap with entity and common types?

Reply:
I only said it's less clear what to do about the extension types overlap, and I would say my mind is not made up yet.

Note that requiring a disambiguation prefix where none used to be required (eg because we added a `date` extension type) is also still a breaking change for user's schema